src/Type/Php/TriggerErrorDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,37 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\ConstantScalarType;
+use PHPStan\Type\NeverType;
+use PHPStan\Type\Type;
+
+class TriggerErrorDynamicReturnTypeExtension implements \PHPStan\Type\DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'trigger_error';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if (count($functionCall->getArgs()) < 2) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		$errorType = $scope->getType($functionCall->getArgs()[1]->value);
+		if ($errorType instanceof ConstantScalarType) {
+			if ($errorType->getValue() === E_USER_ERROR) {
+				return new NeverType();
```

Comment:
We should also test that this behaves as an actual terminating call. Try this:

```php
function foo(?string $s): void
{
    if ($s === null) {
        trigger_error('foo', E_USER_ERROR);
    }
    assertType('string', $s);
}
```

Reply:
hmm interessting.. you are right, this case is failling atm.

I had the impression that phpstan would handle a `NeverType` returned from a DynamicFunctionReturnTypeExtension automatically as a terminating call. wouldn't that make sense, instead of adding such logic case by case?

Reply:
You need `return new NeverType(true);`

Reply:
already fixed :-). thx.